# Goal:
Aim to clean and analyze the data of used car listing.

In [1]:
import pandas as pd
import numpy as np
import chardet
import re


In [2]:
autos= open("autos.csv", "rb")
data=autos.read()
autos.close() 

encoding=chardet.detect(data)['encoding']
autos=pd.read_csv("autos.csv", encoding=encoding, index_col=0, parse_dates= False )

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [5]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

##### Some columns are missing data and will need to be cleaned (eg vehicle type)
different fuel types will effect the care. Seems lise this is a years. 20 columns by 5000 rows.  1 row is the header.


In [6]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [7]:
autos.rename({'yearOfRegistration': "registration_year"}, axis=1).rename({'monthOfRegistration': "registration_month"}, axis=1).rename({'notRepairedDamage': "unrepaired_damage"}, axis=1).rename({'dateCreated': "ad_created"}, axis=1).rename({'offerType': "Offer_type"}, axis=1).rename({'dateCreated': "ad_created"}, axis=1)








dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller Offer_type    price   abtest vehicleType  registration_year  \
0      privat    Angebot   $5,000  control         bus               2004   
1      privat    Angebot   $8,500  control   limousine               1997   
2      privat    Angebot   $8,990     test   limousine               2009   
3      privat    Angebot   $4,350  control  kleinwagen               2007   
4      privat    Angebot   $1,350     test       kombi               2003   
...       ...        ...      ...      ...         ...                ...   
49995  privat    Angebot  $24,900  control   limousine               2011   
49996  privat    Angebot   $1,980  control      cabrio               1996   
49997  privat    Angebot  $13,200     test      cabrio               2014   
49998  privat    Angebot  $22,900  control       kombi               2013   
49999  privat    Angebot   $1,250  control   limousine               1996   

         gearbox  powerPS   model   odometer  registration_month fuelType  \
0        manuell      158  andere  150,000km                   3      lpg   
1      automatik      286     7er  150,000km                   6   benzin   
2        manuell      102    golf   70,000km                   7   benzin   
3      automatik       71  fortwo   70,000km                   6   benzin   
4        manuell        0   focus  150,000km                   7   benzin   
...          ...      ...     ...        ...                 ...      ...   
49995  automatik      239      q5  100,000km                   1   diesel   
49996    manuell       75   astra  150,000km                   5   benzin   
49997  automatik       69     500    5,000km                  11   benzin   
49998    manuell      150      a3   40,000km                  11   diesel   
49999    manuell      101  vectra  150,000km                   1   benzin   

            brand unrepaired_damage           ad_created  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
49995       8213

In [8]:

def camel_to_snake(string):
    string = re.sub('(.)([A-Z][a-z]+)', r'\1_\2', string)
    string = re.sub('(.)([0-9]+)', r'\1_\2', string)
    return re.sub('([a-z0-9])([A-Z])', r'\1_\2', string).lower()



In [9]:
new_col=[]
for each in autos.columns:
    new=camel_to_snake(each)
    new_col.append(new)
autos.columns=new_col
print(autos.columns)
    

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'year_of_registration', 'gearbox', 'power_ps', 'model',
       'odometer', 'month_of_registration', 'fuel_type', 'brand',
       'not_repaired_damage', 'date_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


In [10]:
autos.rename({'year_of_registration': "registration_year"}, axis=1).rename({'month_of_registration': "registration_month"}, axis=1).rename({'not_repaired_damage': "unrepaired_damage"}, axis=1).rename({'date_created': "ad_created"}, axis=1).rename({'date_created': "ad_created"}, axis=1)


date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offer_type    price   abtest vehicle_type  registration_year  \
0      privat    Angebot   $5,000  control          bus               2004   
1      privat    Angebot   $8,500  control    limousine               1997   
2      privat    Angebot   $8,990     test    limousine               2009   
3      privat    Angebot   $4,350  control   kleinwagen               2007   
4      privat    Angebot   $1,350     test        kombi               2003   
...       ...        ...      ...      ...          ...                ...   
49995  privat    Angebot  $24,900  control    limousine               2011   
49996  privat    Angebot   $1,980  control       cabrio               1996   
49997  privat    Angebot  $13,200     test       cabrio               2014   
49998  privat    Angebot  $22,900  control        kombi               2013   
49999  privat    Angebot   $1,250  control    limousine               1996   

         gearbox  power_ps   model   odometer  registration_month fuel_type  \
0        manuell       158  andere  150,000km                   3       lpg   
1      automatik       286     7er  150,000km                   6    benzin   
2        manuell       102    golf   70,000km                   7    benzin   
3      automatik        71  fortwo   70,000km                   6    benzin   
4        manuell         0   focus  150,000km                   7    benzin   
...          ...       ...     ...        ...                 ...       ...   
49995  automatik       239      q5  100,000km                   1    diesel   
49996    manuell        75   astra  150,000km                   5    benzin   
49997  automatik        69     500    5,000km                  11    benzin   
49998    manuell       150      a3   40,000km                  11    diesel   
49999    manuell       101  vectra  150,000km                   1    benzin   

            brand unrepaired_damage           ad_created  nr_of_pictures  \
0         peugeot              nein  2016-03-26 00:00:00               0   
1             bmw              nein  2016-04-04 00:00:00               0   
2      volkswagen              nein  2016-03-26 00:00:00               0   
3           smart              nein  2016-03-12 00:00:00               0   
4            ford              nein  2016-04-01 00:00:00               0   
...           ...               ...                  ...             ...   
49995        audi              nein  2016-03-27 00:00:00               0   
49996        opel              nein  2016-03-28 00:00:00               0   
49997        fiat              nein  2016-04-02 00:00:00               0   
49998        audi              nein  2016-03-08 00:00:00               0   
49999        opel              nein  2016-03-13 00:00:00               0   

       postal_code            last_seen  
0            79588  2016-04-06 06:45:54  
1            71034  2016-04-06 14:45:08  
2            35394  2016-04-06 20:15:37  
3            33729  2016-03-15 03:16:28  
4            39218  2016-04-01 1

In [11]:
print(autos.columns)

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'year_of_registration', 'gearbox', 'power_ps', 'model',
       'odometer', 'month_of_registration', 'fuel_type', 'brand',
       'not_repaired_damage', 'date_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


In [12]:
autos=autos.rename({"year_of_registration": "registration_year"}, axis=1).rename({"month_of_registration": "registration_month"}, axis=1).rename({"not_repaired_damage": "unrepaired_damage"}, axis=1).rename({"date_created": "ad_created"}, axis=1)

In [13]:
print(autos.columns)

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


In [14]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

##### All column header names where change snakecase and more clear header titles

In [15]:
autos.describe(include="all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-10 15:36:24  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

seller, offertype, nm_of_pictures - not needed
unrepared damage- neim - do we need it, 
odometer - 32424 - is it useful?, 
gearbox - almost all manuel but seems import


int - price, odometer
date:date_crawled, ad_created, last_seen, registration_month, registation_year

In [16]:
autos["price"].unique

<bound method Series.unique of 0         $5,000
1         $8,500
2         $8,990
3         $4,350
4         $1,350
          ...   
49995    $24,900
49996     $1,980
49997    $13,200
49998    $22,900
49999     $1,250
Name: price, Length: 50000, dtype: object>

In [17]:
autos["price"] = autos["price"].str.replace("$", "").str.replace(",","").astype(int)



In [18]:
autos['price'].unique

<bound method Series.unique of 0         5000
1         8500
2         8990
3         4350
4         1350
         ...  
49995    24900
49996     1980
49997    13200
49998    22900
49999     1250
Name: price, Length: 50000, dtype: int32>

In [19]:
autos=autos.rename({"price":"price_$"}, axis=1)

In [20]:
autos.columns


Index(['date_crawled', 'name', 'seller', 'offer_type', 'price_$', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [21]:
print(autos["odometer"].unique)

<bound method Series.unique of 0        150,000km
1        150,000km
2         70,000km
3         70,000km
4        150,000km
           ...    
49995    100,000km
49996    150,000km
49997      5,000km
49998     40,000km
49999    150,000km
Name: odometer, Length: 50000, dtype: object>


In [22]:
autos["odometer"] = autos["odometer"].str.replace("km", "").str.replace(",", "").astype(int)

In [23]:
autos["odometer"].unique

<bound method Series.unique of 0        150000
1        150000
2         70000
3         70000
4        150000
          ...  
49995    100000
49996    150000
49997      5000
49998     40000
49999    150000
Name: odometer, Length: 50000, dtype: int32>

In [24]:
autos= autos.rename({"odometer": "odometer_km"}, axis=1)



In [25]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price_$', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [26]:
print(autos["odometer_km"].max())
print(autos["odometer_km"].min())


150000
5000


In [27]:
autos["odometer_km"].unique().shape


(13,)

In [28]:
autos["odometer_km"].describe()


count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [29]:
autos["odometer_km"].value_counts().sort_index(ascending=False)


150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

In [30]:
#added a new column as later cell wanted average with 150000 km
autos['odometer_km<150'] = autos["odometer_km"]
autos["odometer_km"].value_counts().sort_index(ascending=False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

In [31]:
autos[["odometer_km", "odometer_km<150"]]

odometer_km  odometer_km<150
0           150000           150000
1           150000           150000
2            70000            70000
3            70000            70000
4           150000           150000
...            ...              ...
49995       100000           100000
49996       150000           150000
49997         5000             5000
49998        40000            40000
49999       150000           150000

[50000 rows x 2 columns]

In [32]:
print(autos.head(20))

           date_crawled                                               name  \
0   2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1   2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2   2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3   2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4   2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5   2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6   2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7   2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8   2016-03-22 16:51:34                                         Seat_Arosa   
9   2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10  2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11  2016-03-16 18:45:34                         Mercedes_A140_Mo

In [33]:
autos.loc[autos["odometer_km<150"] == 150000, "odometer_km<150"] = np.nan


#autos.loc[((autos["odometer_km<150"]<5000) | (autos["odometer_km<150"]>125000)), "odometer_km<150"] = np.nan
autos["odometer_km<150"].describe()
#autos["odometer_km<150"].value_counts().sort_index(ascending=False)
print(autos["odometer_km<150"].max())
print(autos["odometer_km"].max())
autos.columns

125000.0
150000


Index(['date_crawled', 'name', 'seller', 'offer_type', 'price_$', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen', 'odometer_km<150'],
      dtype='object')

##### We took the average of km without the 150,000km to see the average of all km not including 150,000km.  Excluding 150,000km, the min is 5000km, max is 12,500km  and the average 80,964.66.  64.8% is 150,000km

In [34]:
autos["odometer_km"].value_counts().sort_index(ascending=False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

In [35]:
autos["odometer_km<150"].describe()

count     17576.000000
mean      80964.667729
std       38349.570563
min        5000.000000
25%       50000.000000
50%       90000.000000
75%      125000.000000
max      125000.000000
Name: odometer_km<150, dtype: float64

In [36]:
autos["odometer_km<150"].value_counts().sort_index(ascending=False)

125000.0    5170
100000.0    2169
90000.0     1757
80000.0     1436
70000.0     1230
60000.0     1164
50000.0     1027
40000.0      819
30000.0      789
20000.0      784
10000.0      264
5000.0       967
Name: odometer_km<150, dtype: int64

In [37]:
autos["price_$"].min()


0

In [38]:
autos["price_$"].max()

99999999

In [39]:
autos["price_$"].unique().shape

(2357,)

In [40]:
autos["price_$"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_$, dtype: float64

In [41]:
autos["price_$"].value_counts().sort_index()

0           1421
1            156
2              3
3              1
5              2
            ... 
10000000       1
11111111       2
12345678       3
27322222       1
99999999       1
Name: price_$, Length: 2357, dtype: int64

In [42]:
autos=autos[(autos["price_$"].between(1000,150000))]
autos["price_$"].describe()

count     38606.000000
mean       7145.409625
std        8294.701267
min        1000.000000
25%        2200.000000
50%        4300.000000
75%        8950.000000
max      145000.000000
Name: price_$, dtype: float64

In [43]:
autos.describe(include="all")

date_crawled      name  seller offer_type        price_$  \
count                 38606     38606   38606      38606   38606.000000   
unique                37538     29729       1          1            NaN   
top     2016-03-19 17:36:18  BMW_320i  privat    Angebot            NaN   
freq                      3        67   38606      38606            NaN   
mean                    NaN       NaN     NaN        NaN    7145.409625   
std                     NaN       NaN     NaN        NaN    8294.701267   
min                     NaN       NaN     NaN        NaN    1000.000000   
25%                     NaN       NaN     NaN        NaN    2200.000000   
50%                     NaN       NaN     NaN        NaN    4300.000000   
75%                     NaN       NaN     NaN        NaN    8950.000000   
max                     NaN       NaN     NaN        NaN  145000.000000   

       abtest vehicle_type  registration_year  gearbox      power_ps  ...  \
count   38606        35851       38606.000000    37174  38606.000000  ...   
unique      2            8                NaN        2           NaN  ...   
top      test    limousine                NaN  manuell           NaN  ...   
freq    19928        10544                NaN    27810           NaN  ...   
mean      NaN          NaN        2005.677796      NaN    128.770761  ...   
std       NaN          NaN          86.707289      NaN    215.825108  ...   
min       NaN          NaN        1000.000000      NaN      0.000000  ...   
25%       NaN          NaN        2001.000000      NaN     80.000000  ...   
50%       NaN          NaN        2005.000000      NaN    116.000000  ...   
75%       NaN          NaN        2009.000000      NaN    160.000000  ...   
max       NaN          NaN        9999.000000      NaN  17700.000000  ...   

          odometer_km  registration_month  fuel_type       brand  \
count    38606.000000        38606.000000      36251       38606   
unique            NaN                 NaN          7          40   
top               NaN                 NaN     benzin  volkswagen   
freq              NaN                 NaN      21892        8220   
mean    122827.021706            5.998990        NaN         NaN   
std      40743.465957            3.591291        NaN         NaN   
min       5000.000000            0.000000        NaN         NaN   
25%     100000.000000            3.000000        NaN         NaN   
50%     150000.000000            6.000000        NaN         NaN   
75%     150000.000000            9.000000        NaN         NaN   
max     150000.000000           12.000000        NaN         NaN   

       unrepaired_damage           ad_created nr_of_pictures   postal_code  \
count              32836                38606        38606.0  38606.000000   
unique                 2                   74            NaN           NaN   
top                 nein  2016-04-03 00:00:00            NaN           NaN   
freq               30452                 1523            NaN           NaN   
mean                 NaN                  NaN            0.0  51685.377817   
std                  NaN                  NaN            0.0  25689.297008   
min                  NaN                  NaN            0.0   1067.000000   
25%                  NaN                  NaN            0.0  31205.250000   
50%                  NaN                  NaN            0.0  50829.000000   
75%                  NaN                  NaN            0.0  72529.500000   
max                  NaN                  NaN            0.0  99998.000000   

                  last_seen odometer_km<150  
count                 38606    15293.000000  
unique                31208             NaN  
top     2016-04-07 03:16:17             NaN  
freq                      7             NaN  
mean                    NaN    81403.910286  
std                     NaN    36730.560301  
min                     NaN     5000.000000  
25%                     NaN    50000.000000  
50%                  

In [44]:
autos["date_crawled"].str.split().str[0].value_counts(dropna=False).sort_index()


2016-03-05     985
2016-03-06     536
2016-03-07    1357
2016-03-08    1259
2016-03-09    1254
2016-03-10    1287
2016-03-11    1267
2016-03-12    1444
2016-03-13     618
2016-03-14    1415
2016-03-15    1299
2016-03-16    1122
2016-03-17    1176
2016-03-18     495
2016-03-19    1357
2016-03-20    1473
2016-03-21    1439
2016-03-22    1254
2016-03-23    1243
2016-03-24    1120
2016-03-25    1179
2016-03-26    1279
2016-03-27    1211
2016-03-28    1364
2016-03-29    1312
2016-03-30    1275
2016-03-31    1213
2016-04-01    1337
2016-04-02    1402
2016-04-03    1511
2016-04-04    1424
2016-04-05     515
2016-04-06     126
2016-04-07      58
Name: date_crawled, dtype: int64

The date_crawled is between date 2016-03-05 thru 2016-04-07 meaning all used cars have been entered in just over a month. The time was taken out as it was not significant for data purposes.

In [45]:
var=autos["ad_created"].value_counts(dropna=False).sort_index()
print(var.tail(50))

2016-02-18 00:00:00       2
2016-02-19 00:00:00       3
2016-02-20 00:00:00       1
2016-02-21 00:00:00       2
2016-02-22 00:00:00       1
2016-02-23 00:00:00       4
2016-02-24 00:00:00       2
2016-02-25 00:00:00       3
2016-02-26 00:00:00       1
2016-02-27 00:00:00       4
2016-02-28 00:00:00      10
2016-02-29 00:00:00       7
2016-03-01 00:00:00       4
2016-03-02 00:00:00       5
2016-03-03 00:00:00      33
2016-03-04 00:00:00      59
2016-03-05 00:00:00     888
2016-03-06 00:00:00     585
2016-03-07 00:00:00    1306
2016-03-08 00:00:00    1257
2016-03-09 00:00:00    1261
2016-03-10 00:00:00    1274
2016-03-11 00:00:00    1277
2016-03-12 00:00:00    1434
2016-03-13 00:00:00     682
2016-03-14 00:00:00    1351
2016-03-15 00:00:00    1292
2016-03-16 00:00:00    1144
2016-03-17 00:00:00    1164
2016-03-18 00:00:00     512
2016-03-19 00:00:00    1316
2016-03-20 00:00:00    1477
2016-03-21 00:00:00    1450
2016-03-22 00:00:00    1246
2016-03-23 00:00:00    1233
2016-03-24 00:00:00 

Although there were ads from 2016-02-18, the majority was from 2016-03-03 thru 2016-04-07.

In [46]:
autos["last_seen"].str.split().str[0].value_counts(dropna=False).sort_index()

2016-03-05      42
2016-03-06     138
2016-03-07     176
2016-03-08     241
2016-03-09     343
2016-03-10     379
2016-03-11     453
2016-03-12     857
2016-03-13     324
2016-03-14     463
2016-03-15     579
2016-03-16     596
2016-03-17    1019
2016-03-18     284
2016-03-19     564
2016-03-20     764
2016-03-21     759
2016-03-22     803
2016-03-23     692
2016-03-24     716
2016-03-25     686
2016-03-26     620
2016-03-27     543
2016-03-28     749
2016-03-29     802
2016-03-30     906
2016-03-31     878
2016-04-01     896
2016-04-02     961
2016-04-03     943
2016-04-04     903
2016-04-05    5060
2016-04-06    9063
2016-04-07    5404
Name: last_seen, dtype: int64

The last seen dates are between 2016-03-05 thru 2016-04-07 with the highest last seen on 2016-04-06, viewed 11,050.  All other dates last seen was in the tens to thousands.  The time was taken out as it was not significant for data purposes.

In [47]:
autos["registration_year"].describe()


count    38606.000000
mean      2005.677796
std         86.707289
min       1000.000000
25%       2001.000000
50%       2005.000000
75%       2009.000000
max       9999.000000
Name: registration_year, dtype: float64

In [48]:
autos[autos["registration_year"].between(1997,2008)].describe()
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price_$', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen', 'odometer_km<150'],
      dtype='object')

The registration_year has some incorrect dates.  The dates should be between 1997 and 2008 all other information is irrelevant.  The average is 2002.

The range of years was choosen by analyzing the 25% and 75% and adding a few years to the data.  After the describe was rerun to see if the information ran more accurately.  The infomation now between min, 25%,50%, 75%, and max are in 3 year intervals. I then tested it by seeing the difference with the years 1990- 2016 and noticed the data 75% and 25% information did not average in correctly. 

In [49]:
autos=autos[autos["registration_year"].between(1997,2008)]

In [50]:
autos["registration_year"].describe()

count    24974.000000
mean      2003.235725
std          3.124374
min       1997.000000
25%       2001.000000
50%       2004.000000
75%       2006.000000
max       2008.000000
Name: registration_year, dtype: float64

In [51]:
autos["registration_year"].value_counts(normalize=True).sort_index()

1997    0.031513
1998    0.051333
1999    0.069032
2000    0.080083
2001    0.082646
2002    0.085489
2003    0.099183
2004    0.104429
2005    0.111516
2006    0.106150
2007    0.090414
2008    0.088212
Name: registration_year, dtype: float64

The information above gives us a percentage of cars between the years 1997 and 2008.  All the years are around the same amount of %, with 2000 being the highest at 10.5% and 1997 being the lowest at 6.4%

In [52]:
autos["brand"].unique()

array(['peugeot', 'bmw', 'smart', 'ford', 'chrysler', 'volkswagen',
       'audi', 'renault', 'mazda', 'porsche', 'mini', 'mercedes_benz',
       'opel', 'saab', 'volvo', 'seat', 'nissan', 'skoda',
       'sonstige_autos', 'mitsubishi', 'toyota', 'fiat', 'hyundai',
       'honda', 'suzuki', 'kia', 'chevrolet', 'jaguar', 'subaru',
       'land_rover', 'jeep', 'alfa_romeo', 'rover', 'citroen', 'daihatsu',
       'daewoo', 'dacia', 'lancia', 'lada'], dtype=object)

In [53]:
autos["brand"].value_counts(normalize=True, dropna=False).sort_values()

lada              0.000480
lancia            0.000480
rover             0.000961
dacia             0.001041
daewoo            0.001361
saab              0.001522
jaguar            0.001642
subaru            0.002002
land_rover        0.002322
daihatsu          0.002362
jeep              0.002362
chrysler          0.004004
chevrolet         0.004324
suzuki            0.005646
porsche           0.005846
sonstige_autos    0.006607
kia               0.006847
honda             0.006887
alfa_romeo        0.007007
mini              0.007768
mitsubishi        0.007808
hyundai           0.008489
volvo             0.009410
nissan            0.013454
citroen           0.013934
toyota            0.014695
skoda             0.016537
mazda             0.016657
seat              0.016657
smart             0.017939
fiat              0.018980
peugeot           0.032354
renault           0.042004
ford              0.061023
audi              0.094378
opel              0.098663
mercedes_benz     0.113118
b

There is a varitey of cars.  For analyzing purposes I will be focusin gon 2% and higher.

In [54]:
autos["brand"].value_counts( dropna=False).sort_values().tail(9)

fiat              474
peugeot           808
renault          1049
ford             1524
audi             2357
opel             2464
mercedes_benz    2825
bmw              3270
volkswagen       5032
Name: brand, dtype: int64

In [55]:
autos=autos[(autos["brand"]=="fiat")|(autos["brand"]=="peugeot")|(autos["brand"]=="renault")|(autos["brand"]=="ford")|(autos["brand"]=="audi")|(autos["brand"]=="mercedes_benz")|(autos["brand"]=="bmw")|(autos["brand"]=="opel")|(autos["brand"]=="volkswagen")]


In [56]:
autos["brand"].unique()

array(['peugeot', 'bmw', 'ford', 'volkswagen', 'audi', 'renault',
       'mercedes_benz', 'opel', 'fiat'], dtype=object)

In [57]:
autos["price_$"].describe()

count    19803.000000
mean      4871.687472
std       4362.012351
min       1000.000000
25%       1999.000000
50%       3490.000000
75%       6400.000000
max      85000.000000
Name: price_$, dtype: float64

In [58]:
avg_brand={}


for c in autos["brand"].unique():
    selected_rows=autos[autos["brand"] == c]
    avg_price = sum(selected_rows["price_$"])/len(selected_rows)
    avg_brand[c]=avg_price
print(avg_brand)   

{'peugeot': 2986.70297029703, 'bmw': 6757.6730886850155, 'ford': 3200.4002624671916, 'volkswagen': 4347.577901430843, 'audi': 6575.4212982605, 'renault': 2314.0886558627262, 'mercedes_benz': 6575.489203539823, 'opel': 3028.8283279220777, 'fiat': 2624.9388185654007}


In [59]:
avg_mil={}


for c in autos["brand"].unique():
    selected_rows=autos[autos["brand"] == c]
    avg_price = sum(selected_rows["odometer_km"])/len(selected_rows)
    avg_mil[c]=avg_price
print(avg_mil)   

{'peugeot': 131460.39603960395, 'bmw': 140051.9877675841, 'ford': 133392.38845144358, 'volkswagen': 139154.41176470587, 'audi': 141974.9681798897, 'renault': 133493.80362249762, 'mercedes_benz': 138555.75221238937, 'opel': 134478.49025974027, 'fiat': 126086.49789029536}


In [60]:
autos["odometer_km<150"].describe()

count      4686.000000
mean      97602.432778
std       31837.662977
min        5000.000000
25%       80000.000000
50%      100000.000000
75%      125000.000000
max      125000.000000
Name: odometer_km<150, dtype: float64

In [61]:
print(len(autos["odometer_km<150"]))

19803


In [62]:
avg_mil_not150={}


for c in autos["brand"].unique():
    selected_rows=autos[autos["brand"] == c]
    avg_price = selected_rows["odometer_km<150"].sum()/len(selected_rows)
    avg_mil_not150[c]=avg_price
print(avg_mil_not150)

{'peugeot': 34925.74257425743, 'bmw': 17987.76758409786, 'ford': 28864.82939632546, 'volkswagen': 20066.573926868045, 'audi': 15840.050912176495, 'renault': 33398.474737845565, 'mercedes_benz': 20042.477876106193, 'opel': 30379.464285714286, 'fiat': 45390.29535864979}


In [63]:
avgb = pd.Series(avg_brand)
print(avgb)

peugeot          2986.702970
bmw              6757.673089
ford             3200.400262
volkswagen       4347.577901
audi             6575.421298
renault          2314.088656
mercedes_benz    6575.489204
opel             3028.828328
fiat             2624.938819
dtype: float64


In [64]:
avgm = pd.Series(avg_mil)
print(avgm)

peugeot          131460.396040
bmw              140051.987768
ford             133392.388451
volkswagen       139154.411765
audi             141974.968180
renault          133493.803622
mercedes_benz    138555.752212
opel             134478.490260
fiat             126086.497890
dtype: float64


In [65]:
avgmnot = pd.Series(avg_mil_not150)
print(avgmnot)

peugeot          34925.742574
bmw              17987.767584
ford             28864.829396
volkswagen       20066.573927
audi             15840.050912
renault          33398.474738
mercedes_benz    20042.477876
opel             30379.464286
fiat             45390.295359
dtype: float64


In [66]:
mean = pd.DataFrame(avgm, columns=['odometer_km'])
print(mean)

                 odometer_km
peugeot        131460.396040
bmw            140051.987768
ford           133392.388451
volkswagen     139154.411765
audi           141974.968180
renault        133493.803622
mercedes_benz  138555.752212
opel           134478.490260
fiat           126086.497890


In [67]:
mean['odometer<150'] = avgmnot
print(mean)

                 odometer_km  odometer<150
peugeot        131460.396040  34925.742574
bmw            140051.987768  17987.767584
ford           133392.388451  28864.829396
volkswagen     139154.411765  20066.573927
audi           141974.968180  15840.050912
renault        133493.803622  33398.474738
mercedes_benz  138555.752212  20042.477876
opel           134478.490260  30379.464286
fiat           126086.497890  45390.295359


In [68]:
mean['price_$'] = avgb
print(mean)

                 odometer_km  odometer<150      price_$
peugeot        131460.396040  34925.742574  2986.702970
bmw            140051.987768  17987.767584  6757.673089
ford           133392.388451  28864.829396  3200.400262
volkswagen     139154.411765  20066.573927  4347.577901
audi           141974.968180  15840.050912  6575.421298
renault        133493.803622  33398.474738  2314.088656
mercedes_benz  138555.752212  20042.477876  6575.489204
opel           134478.490260  30379.464286  3028.828328
fiat           126086.497890  45390.295359  2624.938819


In [69]:
autos["seller"].unique()

array(['privat'], dtype=object)

In [70]:
autos["seller"] = autos["seller"].str.replace("privat", "private")

In [71]:
autos["offer_type"].unique()

array(['Angebot'], dtype=object)

In [72]:
autos["offer_type"] = autos["offer_type"].str.replace("Angebot", "offer")

In [73]:
autos["gearbox"] = autos["gearbox"].str.replace("manuell", "manuel").str.replace("automatik", "automatic")
autos["gearbox"].unique()

array(['manuel', 'automatic', nan], dtype=object)

In [74]:

autos["vehicle_type"] = autos["vehicle_type"].str.replace("kombi", "station wagon").str.replace("kleinwagen", "small car").str.replace("cabrio", "convertible").str.replace("andere", "other")
autos["vehicle_type"].unique()

array(['bus', 'limousine', 'station wagon', 'small car', 'convertible',
       nan, 'coupe', 'suv', 'other'], dtype=object)

In [75]:

autos["unrepaired_damage"] = autos["unrepaired_damage"].str.replace("nein", "no").str.replace("ja", "yes")
autos["unrepaired_damage"].unique()

array(['no', nan, 'yes'], dtype=object)

In [76]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price_$', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen', 'odometer_km<150'],
      dtype='object')

In [77]:
autos['date_crawled'].unique

<bound method Series.unique of 0        2016-03-26 17:47:46
1        2016-04-04 13:38:56
4        2016-04-01 14:38:50
7        2016-03-16 18:55:19
13       2016-03-23 10:48:32
                ...         
49977    2016-03-31 18:37:18
49981    2016-03-15 09:38:21
49988    2016-03-28 19:49:51
49993    2016-03-15 18:47:35
49994    2016-03-22 17:36:42
Name: date_crawled, Length: 19803, dtype: object>

In [78]:
autos["date_crawled"]=autos["date_crawled"].str.split().str[0]
autos['date_crawled'].unique

<bound method Series.unique of 0        2016-03-26
1        2016-04-04
4        2016-04-01
7        2016-03-16
13       2016-03-23
            ...    
49977    2016-03-31
49981    2016-03-15
49988    2016-03-28
49993    2016-03-15
49994    2016-03-22
Name: date_crawled, Length: 19803, dtype: object>

In [79]:
autos["date_crawled"] = autos["date_crawled"].str.replace("-", "").astype(int)
autos['date_crawled'].unique

<bound method Series.unique of 0        20160326
1        20160404
4        20160401
7        20160316
13       20160323
           ...   
49977    20160331
49981    20160315
49988    20160328
49993    20160315
49994    20160322
Name: date_crawled, Length: 19803, dtype: int32>

In [80]:
autos['ad_created'].unique

<bound method Series.unique of 0        2016-03-26 00:00:00
1        2016-04-04 00:00:00
4        2016-04-01 00:00:00
7        2016-03-16 00:00:00
13       2016-03-23 00:00:00
                ...         
49977    2016-03-31 00:00:00
49981    2016-03-15 00:00:00
49988    2016-03-28 00:00:00
49993    2016-03-15 00:00:00
49994    2016-03-22 00:00:00
Name: ad_created, Length: 19803, dtype: object>

In [81]:
autos["ad_created"] = autos["ad_created"].str.replace("-", "").str.replace(" 00:00:00", "").astype(int)
autos['date_crawled'].unique

<bound method Series.unique of 0        20160326
1        20160404
4        20160401
7        20160316
13       20160323
           ...   
49977    20160331
49981    20160315
49988    20160328
49993    20160315
49994    20160322
Name: date_crawled, Length: 19803, dtype: int32>

In [82]:
autos['last_seen'].unique

<bound method Series.unique of 0        2016-04-06 06:45:54
1        2016-04-06 14:45:08
4        2016-04-01 14:38:50
7        2016-04-07 03:17:32
13       2016-04-01 14:17:13
                ...         
49977    2016-04-06 12:16:11
49981    2016-04-05 17:21:46
49988    2016-04-07 02:17:40
49993    2016-04-06 19:46:53
49994    2016-04-06 08:16:39
Name: last_seen, Length: 19803, dtype: object>

In [83]:
autos["last_seen"]=autos["last_seen"].str.split().str[0]
autos["last_seen"] = autos["last_seen"].str.replace("-", "").astype(int)
autos['last_seen'].unique

<bound method Series.unique of 0        20160406
1        20160406
4        20160401
7        20160407
13       20160401
           ...   
49977    20160406
49981    20160405
49988    20160407
49993    20160406
49994    20160406
Name: last_seen, Length: 19803, dtype: int32>

In [84]:
autos["name"].head(50)

0                       Peugeot_807_160_NAVTECH_ON_BOARD
1             BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik
4      Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...
7                                   Golf_IV_1.9_TDI_90PS
13                                    Audi_A3_1.6_tuning
14     Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...
20     Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...
29     Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...
31                             Noch_gut_erhaltenen_C_320
33     Audi_A3_2.0_TDI_Sportback_DPF_Kupplung_defekt_...
34                               Bmw_530d__navi_tuv03/18
35                                           Ford_Mondeo
36             Ford_Mondeo_2.0_Turnier_TDCi_DPF_Titanium
38                           Volkswagen_Golf_1.4_Special
39                    Mercedes_C_180_Kombi__zu_verkaufen
40     VW_Golf__4_Cabrio_2.0_GTI_16V___Leder_MFA_Alus...
41     Passat_3b_1.9_TDI_Highline__angemeldet_mit_tuv...
46                             

In [85]:
re
autos.head()

date_crawled                                               name   seller  \
0       20160326                   Peugeot_807_160_NAVTECH_ON_BOARD  private   
1       20160404         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  private   
4       20160401  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...  private   
7       20160316                               Golf_IV_1.9_TDI_90PS  private   
13      20160323                                 Audi_A3_1.6_tuning  private   

   offer_type  price_$   abtest   vehicle_type  registration_year    gearbox  \
0       offer     5000  control            bus               2004     manuel   
1       offer     8500  control      limousine               1997  automatic   
4       offer     1350     test  station wagon               2003     manuel   
7       offer     1990  control      limousine               1998     manuel   
13      offer     1350  control      limousine               1999     manuel   

    power_ps   model  odometer_km  registration_month fuel_type       brand  \
0        158  andere       150000                   3       lpg     peugeot   
1        286     7er       150000                   6    benzin         bmw   
4          0   focus       150000                   7    benzin        ford   
7         90    golf       150000                  12    diesel  volkswagen   
13       101      a3       150000                  11    benzin        audi   

   unrepaired_damage  ad_created  nr_of_pictures  postal_code  last_seen  \
0                 no    20160326               0        79588   20160406   
1                 no    20160404               0        71034   20160406   
4                 no    20160401               0        39218   20160401   
7                 no    20160316               0        53474   20160407   
13                no    20160323               0        12043   20160401   

    odometer_km<150  
0               NaN  
1               NaN  
4               NaN  
7               NaN  
13              NaN

In [86]:
autos["brand_model"]=autos["brand"]+ " " +autos["model"]
print(autos["brand_model"].head())

0      peugeot andere
1             bmw 7er
4          ford focus
7     volkswagen golf
13            audi a3
Name: brand_model, dtype: object


In [87]:
autos["brand_model"].value_counts()

volkswagen golf      1767
bmw 3er              1704
volkswagen passat     876
audi a4               822
bmw 5er               811
                     ... 
opel insignia           2
volkswagen kaefer       2
fiat croma              1
audi q5                 1
audi 100                1
Name: brand_model, Length: 111, dtype: int64

In [88]:
autos["odometer_km"].value_counts()

150000    15117
125000     2077
100000      753
90000       564
80000       388
70000       247
60000       190
5000        164
50000       113
20000        64
30000        60
40000        58
10000         8
Name: odometer_km, dtype: int64

In [94]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price_$', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer_km', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen', 'odometer_km<150', 'brand_model', 'odometer_km=150',
       'odometer_km=125', 'odometer_km100-80', 'odometer_km70-50',
       'odometer_km<50'],
      dtype='object')

In [112]:
autos['odometer_mean'] = autos["odometer_km"]
autos.loc[((autos["odometer_mean"] == 80000) | (autos["odometer_mean"] == 100000)), "odometer_mean"] = 90000
autos.loc[((autos["odometer_mean"] == 50000) | (autos["odometer_mean"] == 70000)), "odometer_mean"] = 60000
autos.loc[(autos["odometer_mean"] < 50000), "odometer_mean"] = 20000
autos['odometer_mean'].value_counts()

150000    15117
125000     2077
90000      1705
60000       550
20000       354
Name: odometer_mean, dtype: int64

In [115]:
grouped=autos.groupby('odometer_mean')[['price_$']]
price_odometer=grouped.agg([np.mean], axis=1)
print(price_odometer.sort_index())


#pv_happy_family_stats= happiness2015.pivot_table(values=["Happiness Score", "Family"], index="Region", aggfunc=([np.min,np.max,np.mean]), margins=True)

                   price_$
                      mean
odometer_mean             
20000          6108.468927
60000          7563.743636
90000          6897.308504
125000         5677.571016
150000         4405.592644


In [116]:
grouped=autos.groupby('unrepaired_damage')[['price_$']]
price_damage=grouped.agg([np.mean], axis=1)
print(price_damage.sort_index())

                       price_$
                          mean
unrepaired_damage             
no                 5234.507979
yes                3127.498542
